### PR TITLE
feat(core): add logs for `OrTransport` when trying addresses

### LINKED ISSUE #4072: core: add useful logs to `OrTransport` when trying addresses
## Description

Currently, there are no logs when `OrTransport` tries an address on several transports: https://github.com/libp2p/rust-libp2p/blob/14530af2611b0cb17179f83802d31c0f28222e91/core/src/transport/choice.rs#L49-L89

We should change that to make these cases easier to debug.

## Open questions

- How to best log the name of the transport? Could use https://doc.rust-lang.org/std/any/fn.type_name.html but that can get very verbose.

## Are you planning to do it yourself in a pull request?

No.

### PR DESCRIPTION
**Pull request by tcoratger**
## Description

Currently, when trying addresses in `listen_on` via `OrTransport`, there are no logs to facilitate debugging. This PR corrects that by providing adequate logs via `std::any::type_name` method.

Resolves #4072.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: core/src/transport/choice.rs [hunk 1]
```diff
@@ -51,13 +51,21 @@ where
         id: ListenerId,
         addr: Multiaddr,
     ) -> Result<(), TransportError<Self::Error>> {
+        println!("Trying address on {}...", std::any::type_name::<A>());
```
Comment (thomaseizinger) [inline core/src/transport/choice.rs]: Please use the `log` crate and trace level instead.

Comment (thomaseizinger) [reply core/src/transport/choice.rs]: Lets say something like "attempting to dial {addr} using {typename}".

File: core/src/transport/choice.rs [hunk 2]
```diff
@@ -51,13 +51,21 @@ where
...
         let addr = match self.0.listen_on(id, addr) {
-            Err(TransportError::MultiaddrNotSupported(addr)) => addr,
+            Err(TransportError::MultiaddrNotSupported(addr)) => {
+                println!("Failed on {}", std::any::type_name::<A>());
```
Comment (thomaseizinger) [inline core/src/transport/choice.rs]: Here, I'd like debug level and also a message that includes address and transport.

**Review by thomaseizinger — COMMENTED** [review]
Thanks! Two suggestions :)

File: core/src/transport/choice.rs (hunk 1 above)
Comment (tcoratger) [reply core/src/transport/choice.rs]: Done

File: core/src/transport/choice.rs (hunk 2 above)
Comment (tcoratger) [reply core/src/transport/choice.rs]: Done

**Review by thomaseizinger — COMMENTED** [review]
Thanks for addressing the comments! Did you test this with an application or an example?

I am curious what these log lines end up looking like.

File: core/src/transport/choice.rs [hunk 3]
```diff
@@ -51,13 +52,37 @@ where
...
+        trace!(
+            "Attempting to dial {} using {}",
+            addr,
+            std::any::type_name::<A>()
+        );
         let addr = match self.0.listen_on(id, addr) {
-            Err(TransportError::MultiaddrNotSupported(addr)) => addr,
+            Err(TransportError::MultiaddrNotSupported(addr)) => {
+                debug!(
+                    "Failing to dial {} using {}",
```
Comment (mxinden) [inline core/src/transport/choice.rs]:
```suggestion
                    "Failed to dial {} using {}",
```

Nit: at this point it already failed, right?

File: core/src/transport/choice.rs [hunk 4]
```diff
@@ -51,13 +52,37 @@ where
...
+        trace!(
+            "Attempting to dial {} using {}",
+            addr,
+            std::any::type_name::<B>()
+        );
         let addr = match self.1.listen_on(id, addr) {
-            Err(TransportError::MultiaddrNotSupported(addr)) => addr,
+            Err(TransportError::MultiaddrNotSupported(addr)) => {
+                debug!(
+                    "Failing to dial {} using {}",
```
Comment (mxinden) [inline core/src/transport/choice.rs]:
```suggestion
                    "Failed to dial {} using {}",
```

Nit: at this point it already failed, right?

**Review by mxinden — COMMENTED** [review]
Neat. This will help a lot in the future. Thank you.

Also curious what the output looks like. Mind posting it here?

**Comment by tcoratger** [comment]
@mxinden @thomaseizinger I've tested with the simplest template in `protocols/relay`. Here is an example response in the console:

```shell
[2023-06-30T19:12:43Z TRACE libp2p_core::transport::choice] Attempting to dial /memory/3374154575862174647 using libp2p_relay::priv_client::transport::Transport
[2023-06-30T19:12:43Z DEBUG libp2p_core::transport::choice] Failed to dial /memory/3374154575862174647 using libp2p_relay::priv_client::transport::Transport
[2023-06-30T19:12:43Z TRACE libp2p_core::transport::choice] Attempting to dial /memory/3374154575862174647 using libp2p_core::transport::memory::MemoryTransport
```

**Comment by thomaseizinger** [comment]
Perhaps we should reduce this to just logging the failures to be less noisy?

My guess is that connections succeeding is almost never an issue. It is the rejections that are hard to debug.

Thoughts guys?

**Comment by tcoratger** [comment]
> Perhaps we should reduce this to just logging the failures to be less noisy?
> 
> My guess is that connections succeeding is almost never an issue. It is the rejections that are hard to debug.
> 
> Thoughts guys?

@thomaseizinger As you wish, for me it's not a major problem because the `trace!` represents a level above in terms of logs and therefore if the user indicates a `trace` level of log, he knows that he will have additional information. If it wants failures it can set to `debug`, I thought that was why you favored using `trace` and `debug` separately so the user can choose.

But if you prefer, no worries, I can just remove the two `traces!` :)

**Comment by mxinden** [comment]
Given that the success case is on TRACE level this looks good to me. Though not a strong opinion. Will defer to @thomaseizinger.

Thank you @tcoratger.

**Review by thomaseizinger — DISMISSED** [review]

**Review by mxinden — APPROVED** [review]

**Pull request merged**